File: bindings/python/examples/toggle_line_value.py
#!/usr/bin/env python3
# SPDX-License-Identifier: GPL-2.0-or-later
# SPDX-FileCopyrightText: 2023 Kent Gibson <warthog618@gmail.com>

"""Minimal example of toggling a single line."""

import gpiod
import time

from gpiod.line import Direction, Value


def toggle_value(value):
    if value == Value.INACTIVE:
        return Value.ACTIVE
    return Value.INACTIVE


def toggle_line_value(chip_path, line_offset):
    value_str = {Value.ACTIVE: "Active", Value.INACTIVE: "Inactive"}
    value = Value.ACTIVE

    with gpiod.request_lines(
        chip_path,
        consumer="toggle-line-value",
        config={
            line_offset: gpiod.LineSettings(
                direction=Direction.OUTPUT, output_value=value
            )
        },
    ) as request:
        while True:
            print("{}={}".format(line_offset, value_str[value]))
            time.sleep(1)
            value = toggle_value(value)
            request.set_value(line_offset, value)


if __name__ == "__main__":
    try:
        toggle_line_value("/dev/gpiochip0", 5)
    except OSError as ex:
        print(ex, "\nCustomise the example configuration to suit your situation")

File: bindings/rust/libgpiod/examples/toggle_line_value.rs
// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
// SPDX-FileCopyrightText: 2023 Kent Gibson <warthog618@gmail.com>
//
// Minimal example of toggling a single line.

use core::time::Duration;
use libgpiod::{
    Result,
    chip::Chip,
    line::{Config as LineConfig, Direction, Settings, Value},
    request::Config as ReqConfig,
};
use std::thread::sleep;

fn toggle_value(value: Value) -> Value {
    match value {
        Value::Active => Value::InActive,
        Value::InActive => Value::Active,
    }
}

fn main() -> Result<()> {
    // Example configuration - customize to suit your situation
    let chip_path = "/dev/gpiochip0";
    let line_offset = 5;
    let mut value = Value::Active;

    let mut settings = Settings::new()?;
    settings
        .set_direction(Direction::Output)?
        .set_output_value(value)?;

    let mut lconfig = LineConfig::new()?;
    lconfig.add_line_settings(&[line_offset], settings)?;

    let mut rconfig = ReqConfig::new()?;
    rconfig.set_consumer("toggle-line-value")?;

    let chip = Chip::open(&chip_path)?;
    let mut req = chip.request_lines(Some(&rconfig), &lconfig)?;

    loop {
        println!("{line_offset}={value:?}");
        sleep(Duration::from_secs(1));
        value = toggle_value(value);
        req.set_value(line_offset, value)?;
    }
}
